### ISSUE TITLE
Make the log10 P values in threshold plot more intuitive?

### ISSUE THREAD

**Issue by yueqiw** [JonathanShor/DoubletDetection#102]
Hi,

The log10 P values (X axis) of in the threshold plot are not quite intuitive right now. The numbers are mostly something like (-2.4, -3.5, 11.2, etc). It would be easier to read if they were integers (-2, -3, -4, etc).

**Comment by adamgayoso** [comment]
That's a good point. Thanks for bringing this up.

**Comment by adamgayoso** [comment]
In the meantime you can pass a np array of p-values through the `p_grid` parameter. If `log10==True` (default), then this array should be log10 p-values.

**Comment by yueqiw** [comment]
Thanks!

**Comment by adamgayoso** [comment]
@yueqiw I found a bug so my comment above isn't going to work. We will push an update soon.

@JonathanShor The default now is to use the p-values that appear during the iterations. We can make the plot much faster if we take the min p value seen and set a max to be the default threshold and then check every integer in between (in log space)

**Comment by JonathanShor** [comment]
@adamgayoso Bin by integer? That works for me.

**Comment by adamgayoso** [comment]
@yueqiw I also want to mention that 2.4 will also see a new default `p_thresh` of 1e-7. Were you able to use the threshold plot to pick thresholds?

**Comment by yueqiw** [comment]
@adamgayoso 

I've been using `p_thresh = 0.001, voter_thresh=0.95` for now (with `n_iter=50`). I'm trying to be conservative on filtering out cells and this threshold gives me 2%-4% predicted doublets across different batches. Is there a particular reason for the 1e-7 default? 

> Were you able to use the threshold plot to pick thresholds?

Yes, as below. 
<img src=https://user-images.githubusercontent.com/5046690/41681189-160b391a-7491-11e8-99ab-f7e8362014f9.png width=50%>

**Comment by yueqiw** [comment]
In addition, would you recommend using `new_lib_as = None` or `np.sum`? I found setting `new_lib_as = None` gives me 2x the amount of putative doublets compared to `np.sum`  in some batches, but similar results on the others.

**Comment by adamgayoso** [comment]
The p-value threshold was calibrated using our internal data. For your second question, do you mean `np.max`? `np.sum` should give the same exact results

**Comment by yueqiw** [comment]
@adamgayoso Sorry I meant `np.max`.

**Comment by yueqiw** [comment]
@adamgayoso 

> It's possible that the difference in result between np.max (if that's what you meant) and None is the detection of large library size singlets of a cell type that represents a large fraction of the data.

Prior to running DoubletDetection, I have filtered the cells based on number of total UMIs, so cells with abnormally high or low UMIs have been filtered out. Do you run DoubletDetection before or after QC filtering? I haven't looked at what are the cells that differ between `None` and `np.max`, but I'll look into it.

**Comment by adamgayoso** [comment]
Typically after QC filtering. That said, since everything is relative in the algorithm, it's still possible that if one cell type dominates that sample that cells with many UMIs are targeted by the synthetic doublets.

**Comment by yueqiw** [comment]
> It's possible that the difference in result between np.max (if that's what you meant) and None is the detection of large library size singlets of a cell type that represents a large fraction of the data.

I looked into the cells that differ between `new_lib_as=None` and `np.max`. `None` gives me 7.5% positive cells, and `np.max` 3.6%. Over 95% of positives cells in `np.max` are also positive in `None`. So using `None` gives me an extra set of putative doublets. It does look like these extra positive cells from `None` are just cells with high UMIs (but probably singlets) of dominant cell types. (1) The extra positive cells in `None` do have high nUMI, higher than dataset average, but similar to the `np.max` positive cells. (2) The `np.max` positive cells tend to group together at several places on t-SNE and umap embeddings, whereas the extra positive cells are almost uniformly distributed across the main cell types. So I think they're unlikely to be cross-type doublets.

**Comment by yueqiw** [comment]
@adamgayoso I got better results using a custom normalizer. Now with `new_lib_as=None`, the set of predicted doublets is a stable across a wide parameter range (1.4-1.8% with p value ranging from 0.001 - 1e-7 and voter threshold of 09 and 0.95), and I no longer see positive cells uniformly distributed across the dominant cell types. The result is similar if using `new_lib_as=np.max`.

I think normalization and preprocessing is a pretty important step for doublet detection. Since the key assumption for doublet detection is that doublets tend to cluster together, and that clustering is often heavily influenced by technical factors and normalization steps, I decided to follow a robust preprocessing procedure. Basically I wrapped the Seurat-style preprocessing steps (in Scanpy) as a custom `normalizer` function, with feature selection and regressing out library size. (Following the Scanpy tutorial: https://nbviewer.jupyter.org/github/theislab/scanpy_usage/blob/master/170505_seurat/seurat.ipynb) 

```
import scanpy.api as sc
from scipy.sparse import issparse

def scanpy_normalizer(count_data):
    adata = sc.AnnData(X = count_data)
    if issparse(adata.X):
        adata.obs['n_counts'] = adata.X.sum(axis=1).A1
    else:
        adata.obs['n_counts'] = adata.X.sum(axis=1)
    sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)
    filter_result = sc.pp.filter_genes_dispersion(
        adata.X, min_mean=0.02, max_mean=3, min_disp=0.8)
    adata = adata[:, filter_result.gene_subset]
    sc.pp.log1p(adata)
    sc.pp.regress_out(adata, ['n_counts'], n_jobs=8)
    sc.pp.scale(adata, max_value=10)
    return adata.X
```

Then PCA is run on the scaled expression values of highly variable genes with nUMI regressed out, same as the Seurat pipeline.

**Comment by adamgayoso** [comment]
This is a very good result. I would recommend the more extensive preprocessing over `new_lib_as=np.max`. 

We also have plans to allow a custom clustering algorithm, though many popular ones are based off PhenoGraph.

**Comment by yueqiw** [comment]
Thanks! I'll use this result for for now. I think louvain-based algorithms (PhenoGraph, scanpy, Seurat) should give quite similar results to each other.

**Comment by adamgayoso** [comment]
@yueqiw how large are the individual datasets? 1.4-1.8% seems small, but this is really the inferred cross-type doublet rate.

**Comment by yueqiw** [comment]
@adamgayoso Actually the result I decided to use was 2.5% (The 1.4-1.8% result came from default n_pcs = 30. I reduced the number of PCs to 20, closer to the clustering parameter I use on this dataset) 

I have 4 runs, each with 2000-3000 cells from 10x genomics. But of course, the real doublets should include both cross-type and within-type doublets, and the algorithm is only predicting cross-type doublets. I'm also trying to be conservative on filtering out cells.
